File: solutions/ts/02.ts
function manufactureGifts(
  giftsToProduce: Array<{ toy: string; quantity: number }>
): string[] {
  return giftsToProduce
    .filter((order) => order.quantity > 0)
    .reduce((gifts, order) => {
      return gifts.concat(new Array(order.quantity).fill(order.toy));
    }, [] as string[]);
}

const production1 = [
  { toy: "car", quantity: 3 },
  { toy: "doll", quantity: 1 },
  { toy: "ball", quantity: 2 },
];

const result1 = manufactureGifts(production1);
console.log(result1);
// ['car', 'car', 'car', 'doll', 'ball', 'ball']

const production2 = [
  { toy: "train", quantity: 0 }, // not manufactured
  { toy: "bear", quantity: -2 }, // neither
  { toy: "puzzle", quantity: 1 },
];

const result2 = manufactureGifts(production2);
console.log(result2);
// ['puzzle']

const production3: Array<{ toy: string; quantity: number }> = [];
const result3 = manufactureGifts(production3);
console.log(result3);
// []

/* 
  Score: ★★★★★★★★ (3 bonus stars for solving it today!)
  Points: 160
  ops: 1695
*/

File: solutions/python/02.py
def manufacture_gifts(gifts_to_produce):
    result = []
    for order in gifts_to_produce:
        quantity = order['quantity']
        if isinstance(quantity, int) and quantity > 0:
            result.extend([order['toy']]*order['quantity'])
    return result


production1 = [
    {"toy": "car", "quantity": 3},
    {"toy": "doll", "quantity": 1},
    {"toy": "ball", "quantity": 2},
]

result1 = manufacture_gifts(production1)
print(list(result1))
# ['car', 'car', 'car', 'doll', 'ball', 'ball']

production2 = [
    {"toy": "train", "quantity": 0},  # not manufactured
    {"toy": "bear", "quantity": -2},  # neither
    {"toy": "puzzle", "quantity": 1},
]

result2 = manufacture_gifts(production2)
print(list(result2))
# ['puzzle']

production3 = []
result3 = manufacture_gifts(production3)
print(list(result3))
# []

"""
  Score: ★★★★★★★ (3 bonus stars for solving it today!)
  Points: 140
  ops: 1680
"""
